File: apps/web/fix_storage.py
import os
import re

known_keys = [
    'clinic_users', 'clinic_current_user', 'pageDailyReports', 'kpiTargets',
    'revenueRecords', 'pagePhoneAssignments', 'customerAppointments',
    'monthlyPayrolls', 'staffExpenseClaims', 'attendanceRecords',
    'attendanceRequests', 'approvalNotifications', 'surgicalCareAssignments',
    'mediaDailyReports', 'cskhDailyReports', 'marketingDailyReports',
    'saleOfflineDailyReports'
]

def process_file(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    original_content = content
    changed = False

    for key in known_keys:
        # JSON.parse(localStorage.getItem('key') || '[]')
        pattern1 = re.compile(r"JSON\.parse\(localStorage\.getItem\(['\"`]" + key + r"['\"`]\)\s*\|\|\s*['\"`]\[\]['\"`]\)")
        if pattern1.search(content):
            content = pattern1.sub(f"getStorageItem('{key}', [])", content)
            changed = True
            
        # localStorage.getItem('key')
        pattern2 = re.compile(r"localStorage\.getItem\(['\"`]" + key + r"['\"`]\)")
        if pattern2.search(content):
            content = pattern2.sub(f"getStorageItem('{key}')", content)
            changed = True

        # localStorage.setItem('key', JSON.stringify(var))
        pattern3 = re.compile(r"localStorage\.setItem\(['\"`]" + key + r"['\"`]\s*,\s*JSON\.stringify\((.*?)\)\)")
        if pattern3.search(content):
            content = pattern3.sub(f"setStorageItem('{key}', \\1)", content)
            changed = True

        # localStorage.setItem('key', var)
        pattern4 = re.compile(r"localStorage\.setItem\(['\"`]" + key + r"['\"`]\s*,\s*([^)]+)\)")
        if pattern4.search(content):
            content = pattern4.sub(f"setStorageItem('{key}', \\1)", content)
            changed = True

        # localStorage.removeItem('key')
        pattern5 = re.compile(r"localStorage\.removeItem\(['\"`]" + key + r"['\"`]\)")
        if pattern5.search(content):
            content = pattern5.sub(f"removeStorageItem('{key}')", content)
            changed = True

    if changed:
        if 'getStorageItem' in content or 'setStorageItem' in content or 'removeStorageItem' in content:
            if 'utils/storageStore.js' not in content:
                import_stmt = "import { getStorageItem, setStorageItem, removeStorageItem } from '@/utils/storageStore.js';\n"
                last_import = content.rfind('import ')
                if last_import != -1:
                    end_of_line = content.find('\n', last_import)
                    content = content[:end_of_line + 1] + import_stmt + content[end_of_line + 1:]
                else:
                    content = import_stmt + content
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        print(f"Updated {filepath}")

for root, _, files in os.walk('src'):
    for file in files:
        if file.endswith('.js') or file.endswith('.jsx'):
            process_file(os.path.join(root, file))

File: apps/web/fix_storage.js
const fs = require('fs');
const path = require('path');

const walk = (dir) => {
  let results = [];
  const list = fs.readdirSync(dir);
  list.forEach((file) => {
    file = path.join(dir, file);
    const stat = fs.statSync(file);
    if (stat && stat.isDirectory()) {
      results = results.concat(walk(file));
    } else if (file.endsWith('.js') || file.endsWith('.jsx')) {
      results.push(file);
    }
  });
  return results;
};

const files = walk('src');

files.forEach(file => {
  let content = fs.readFileSync(file, 'utf8');
  let changed = false;

  // We need to add the import if it's not there and we use it
  const needsImport = /localStorage\.(getItem|setItem|removeItem)/.test(content) && !content.includes("utils/storageStore.js");

  // Only replace for STORAGE_KEYS. Actually it's easier to replace all or just specific ones.
  // Actually, replacing all localStorage might break things like theme settings.
  // Let's replace ONLY for known keys.
  const knownKeys = [
    'clinic_users', 'clinic_current_user', 'pageDailyReports', 'kpiTargets',
    'revenueRecords', 'pagePhoneAssignments', 'customerAppointments',
    'monthlyPayrolls', 'staffExpenseClaims', 'attendanceRecords',
    'attendanceRequests', 'approvalNotifications', 'surgicalCareAssignments',
    'mediaDailyReports', 'cskhDailyReports', 'marketingDailyReports',
    'saleOfflineDailyReports'
  ];

  // For now, let's just make a regex to replace JSON.parse(localStorage.getItem('KEY') || '[]')
  // We can just use a regex for all known keys.
  
  knownKeys.forEach(key => {
    const getItemRegex = new RegExp(`JSON\\.parse\\(localStorage\\.getItem\\(['"\`]${key}['"\`]\\)\\s*\\|\\|\\s*['"\`]\\[\\]['"\`]\\)`, 'g');
    if (getItemRegex.test(content)) {
      content = content.replace(getItemRegex, `getStorageItem('${key}', [])`);
      changed = true;
    }

    const getItemNullRegex = new RegExp(`localStorage\\.getItem\\(['"\`]${key}['"\`]\\)`, 'g');
    if (getItemNullRegex.test(content)) {
      content = content.replace(getItemNullRegex, `getStorageItem('${key}')`);
      changed = true;
    }

    const setItemRegex = new RegExp(`localStorage\\.setItem\\(['"\`]${key}['"\`]\\s*,\\s*JSON\\.stringify\\(([^)]+)\\)\\)`, 'g');
    if (setItemRegex.test(content)) {
      content = content.replace(setItemRegex, `setStorageItem('${key}', $1)`);
      changed = true;
    }

    const setItemRawRegex = new RegExp(`localStorage\\.setItem\\(['"\`]${key}['"\`]\\s*,\\s*([^)]+)\\)`, 'g');
    if (setItemRawRegex.test(content)) {
      content = content.replace(setItemRawRegex, `setStorageItem('${key}', $1)`);
      changed = true;
    }
    
    const removeItemRegex = new RegExp(`localStorage\\.removeItem\\(['"\`]${key}['"\`]\\)`, 'g');
    if (removeItemRegex.test(content)) {
      content = content.replace(removeItemRegex, `removeStorageItem('${key}')`);
      changed = true;
    }
  });

  if (changed) {
    if (!content.includes('getStorageItem') && !content.includes('setStorageItem') && !content.includes('removeStorageItem')) {
        // Did not actually inject the words
    } else {
        if (!content.includes('utils/storageStore.js')) {
            const importStmt = "import { getStorageItem, setStorageItem, removeStorageItem } from '@/utils/storageStore.js';\n";
            // Insert after the last import, or at the top
            const lastImportIndex = content.lastIndexOf('import ');
            if (lastImportIndex !== -1) {
                const endOfLine = content.indexOf('\n', lastImportIndex);
                content = content.slice(0, endOfLine + 1) + importStmt + content.slice(endOfLine + 1);
            } else {
                content = importStmt + content;
            }
        }
        fs.writeFileSync(file, content, 'utf8');
        console.log(`Updated ${file}`);
    }
  }
});
